packages/flutter/lib/src/gestures/binding.dart

```diff
@@ -16,6 +17,131 @@ import 'events.dart';
 import 'hit_test.dart';
 import 'pointer_router.dart';
 import 'pointer_signal_resolver.dart';
+import 'resampler.dart';
+
+typedef HandleSampleTimeChangedCallback = void Function();
+
+// Class that handles resampling of touch events for multiple pointer
+// devices.
+//
+// SchedulerBinding's `currentSystemFrameTimeStamp` is used to determine
+// sample time.
+class _Resampler {
+  _Resampler(this._handlePointerEvent, this._handleSampleTimeChanged);
+
+  // Resamplers used to filter incoming pointer events.
+  final Map<int, PointerEventResampler> _resamplers = <int, PointerEventResampler>{};
+
+  // Flag to track if a frame callback has been scheduled.
+  bool _frameCallbackScheduled = false;
+
+  // Current frame time for resampling.
+  Duration _frameTime = Duration.zero;
+
+  // Last sample time and time stamp of last event.
+  //
+  // Only used for debugPrint of resampling margin.
+  Duration _lastSampleTime = Duration.zero;
+  Duration _lastEventTime = Duration.zero;
+
+  // Callback used to handle pointer events.
+  final HandleEventCallback _handlePointerEvent;
+
+  // Callback used to handle sample time changes.
+  final HandleSampleTimeChangedCallback _handleSampleTimeChanged;
+
+  // Enqueue `events` for resampling or dispatch them directly if
+  // not a touch event.
+  void addOrDispatchAll(Queue<PointerEvent> events) {
+    final SchedulerBinding? scheduler = SchedulerBinding.instance;
+    assert(scheduler != null);
+
+    while (events.isNotEmpty) {
+      final PointerEvent event = events.removeFirst();
+
+      // Add touch event to resampler or dispatch pointer event directly.
+      if (event.kind == PointerDeviceKind.touch) {
+        // Save last event time for debugPrint of resampling margin.
+        _lastEventTime = event.timeStamp;
+
+        final PointerEventResampler resampler = _resamplers.putIfAbsent(
+          event.device,
+          () => PointerEventResampler(),
```

Comment:
Is it true that when this constructor is called (i.e., `event.device` key is absent in `_resamplers`), `event` can only be `PointerAddedEvent` or `PointerDownEvent`?

Reply:
No, that would only be true if resampling was enabled before the first pointer event. If you enable resampling after first receiving a down event then we'll need to create the resampler here when processing a move event.